parsec/backend/asgi/rpc.py

```diff
@@ -63,15 +95,173 @@ async def anonymous_api(raw_organization_id: str):
+
+    cmd_rep = await cmd_func(client_ctx, msg)
+    rep = rpc_msgpack_rep(cmd_rep)
+    rep.headers.add("Api-Version", str(client_ctx.api_version))
+    return rep
+
+
+def get_api_version(headers: Headers) -> ApiVersion:
+    """Parse `Api-version` from the HTTP Header and return a compatible `ApiVersion` that can be used by the server & client"""
+
+    try:
+        raw_api_version = request.headers["Api-Version"]
+        client_api_version = ApiVersion.from_str(raw_api_version)
+        settle_api_version, _client_api_version = settle_compatible_versions(
+            ServerHandshake.SUPPORTED_API_VERSIONS, [client_api_version]
+        )
+    except (ValueError, KeyError):
+        abort(400)
+    except IncompatibleAPIVersionsError:
+        abort(400)
+    return settle_api_version
+
+
+async def verify_auth_request(
+    body: bytes, headers: Headers, organization: Organization, backend: BackendApp
+) -> Tuple[User, Device]:
+
+    try:
+        authorization_method = headers["Authorization"]
+    except KeyError:
+        abort(400)
+
+    if authorization_method == "PARSEC-SIGN-ED25519":
+        try:
+            raw_device_id = headers["Author"]
+            raw_timestamp = headers["Timestamp"].encode()
+            raw_signature_b64 = headers["Signature"]
+        except KeyError:
+            abort(400)
+
+        try:
+            signature_bytes = base64.b64decode(raw_signature_b64)
+        except binascii.Error:
+            abort(400)
+
+        try:
+            device_id = DeviceID(base64.b64decode(raw_device_id).decode())
+        except binascii.Error:
+            abort(400)
+        except ValueError:
+            abort(400)
+        return await verify_auth_request_ed25519(
+            body=body,
+            device_id=device_id,
+            raw_timestamp=raw_timestamp,
+            signature=signature_bytes,
+            organization=organization,
+            backend=backend,
+        )
+    else:
+        raise InvalidHeaderValueError("Authorization", authorization_method)
+
+
+async def verify_auth_request_ed25519(
+    body: bytes,
+    device_id: DeviceID,
+    raw_timestamp: bytes,
+    signature: bytes,
+    organization: Organization,
+    backend: BackendApp,
+) -> Tuple[User, Device]:
+    try:
+        user, device = await backend.user.get_user_with_device(
+            organization.organization_id, device_id
+        )
+    except KeyError:
+        raise DeviceNotFoundError(organization.organization_id, device_id)
+    else:
+        if user.revoked_on:
+            raise UserIsRevokedError(organization.organization_id, user.user_id)
+
+    verify_key = device.verify_key
+
+    try:
+        # when the verify failed, `verify()` raise the exception `BadSignatureError`
+        verify_key.verify_with_signature(
+            signature=signature,
+            message=base64.b64encode(str(device_id).encode()) + raw_timestamp + body,
+        )
+    except Exception as e:
```

Comment:
Don't catch global `Exception` but the actual type of exception you know are possible

Reply:
done